crates/biome_parser/CONTRIBUTING.md

```diff
@@ -0,0 +1,382 @@
+it parses the block statement if it is present in the source code and adds a missing marker and an error if not.
+
+Using the above-described rules result in the following implementation for the `if` statement rule.
+
+```rust, ignore
+fn parse_if_statement(p: &mut Parser) -> ParsedSyntax {
+ if !p.at(T![if]) {
+  return Absent;
+ }
+
+ let m = p.start();
+
+ p.expect(T![if]);
+ p.expect(T!['(']);
+ parse_any_expression(p).or_add_diagnostic(p, js_parse_errors::expeced_if_statement);
+ p.expect(T![')']);
+ parse_block_statement(p).or_add_diagnostic(p, js_parse_errors::expected_block_statement);
+// the else block is optional, handle the marker by using `ok`
+ parse_else_clause(p).ok();
+
+ Present(m.complete(p, JS_IF_STATEMENT));
+}
+```
+
+Hold on, what are these *missing* markers? Biome's AST facade uses fixed offsets to retrieve a particular child from a node.
+For example, the 3rd child of the if statement is the condition. However, the condition would become the second element
+if the opening parentheses `(` isn't present in the source text. That's where missing elements come into play.
+
+### Parsing Lists & Error Recovery
+
+Parsing lists is different from parsing single elements with a fixed set of children because it requires looping until
+the parser reaches a terminal token (or the end of the file).
+
+You may remember that `parse_*` methods shouldn't progress parsing if they return `Absent`.
+Not progressing the parser is problematic inside `while` loops because it inevitably results in an infinite loop.
+
+That's why you must do error recovery when parsing lists. Luckily, the parser comes with the infrastructure to make error recovery a piece of cake.
+The general structure for parsing a list is (yes, that's something the parser infrastructure should provide for you):
+
+
+Let's try to parse an array:
+
+```js
+[ 1, 3, 6 ]
+```
+
+We will use  `ParseSeparatedList` in order to achieve that
+
+```rust, ignore
+struct ArrayElementsList;
+
+impl ParseSeparatedList for ArrayElementsList {
+    type ParsedElement = CompletedMarker;
+
+    fn parse_element(&mut self, p: &mut Parser) -> ParsedSyntax<Self::ParsedElement> {
+        parse_array_element(p)
+    }
+
+    fn is_at_list_end(&self, p: &mut Parser) -> bool {
+        p.at_ts(token_set![T![default], T![case], T!['}']])
+    }
+
+    fn recover(
+        &mut self,
+        p: &mut Parser,
+        parsed_element: ParsedSyntax<Self::ParsedElement>,
+    ) -> parser::RecoveryResult {
+        parsed_element.or_recover(
+            p,
+            &ParseRecoveryTokenSet::new(JS_BOGUS_STATEMENT, STMT_RECOVERY_SET),
+            js_parse_error::expected_case,
+        )
+    }
+};
+```
+
+Let's run through this step by step:
+
+```rust, ignore
+parsed_element.or_recover(
+    p,
+    &ParseRecoveryTokenSet::new(JS_BOGUS_STATEMENT, STMT_RECOVERY_SET),
+    js_parse_error::expected_case,
```

Comment:
The docs say we’re parsing an array, but the example seems to be for switch statements?

Reply:
Definitely outdate, I will take care of it later